modules/fundamental/src/analysis/service/mod.rs

```diff
@@ -0,0 +1,564 @@
+use crate::Error;
+use sea_orm::{prelude::ConnectionTrait, Statement};
+use std::collections::HashMap;
+use tracing::instrument;
+use trustify_common::{
+    db::{query::Query, Database, Transactional},
+    model::{Paginated, PaginatedResults},
+};
+
+use petgraph::graph::{Graph, NodeIndex};
+// use petgraph::visit::EdgeRef;
+use itertools::Itertools;
+use petgraph::Direction;
+use regex::Regex;
+use serde::Serialize;
+use std::fmt;
+use std::str::FromStr;
+use trustify_common::purl::Purl;
+use trustify_entity::relationship::Relationship;
+use utoipa::ToSchema;
+
+#[derive(Debug, Clone, PartialEq, Eq, ToSchema, serde::Serialize)]
+pub struct PackageNode {
+    purl: String,
+    name: String,
+    published: String,
+    document_id: String,
+    product_name: String,
+    product_version: String,
+}
+impl fmt::Display for PackageNode {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.name)
+    }
+}
+
+#[derive(Debug, Clone, Serialize)]
+pub struct AdvisoryGraphSummary {
+    pub component: String,
+    pub ancestors: Vec<PackageNode>,
+}
+impl AdvisoryGraphSummary {
+    fn new(component: String, ancestors: Vec<PackageNode>) -> AdvisoryGraphSummary {
+        AdvisoryGraphSummary {
+            component,
+            ancestors,
+        }
+    }
+}
+
+pub struct AnalysisService {
+    db: Database,
+}
+
+#[allow(clippy::too_many_arguments)]
+pub fn add_node(
+    g: &mut Graph<PackageNode, Relationship, petgraph::Directed>,
+    nodes: &mut HashMap<String, NodeIndex>,
+    component_purl: &str,
+    component_name: &str,
+    sbom_published: &str,
+    document_id: &str,
+    product_name: &str,
+    product_version: &str,
+) -> NodeIndex {
+    match nodes.get(&component_purl.to_string()) {
+        Some(&i) => i,
+        None => {
+            let i = g.add_node(PackageNode {
+                purl: component_purl.to_string(),
+                name: component_name.to_string(),
+                published: sbom_published.to_string(),
+                document_id: document_id.to_string(),
+                product_name: product_name.to_string(),
+                product_version: product_version.to_string(),
+            });
+            nodes.insert(component_purl.to_string(), i);
+            i
+        }
+    }
+}
+
+pub fn regex_purl_find(nodes: &HashMap<String, NodeIndex>, search_value: &str) -> Vec<NodeIndex> {
```

Comment:
Regexps can be quite costly sometimes. While the `Pattern` trait seems still to be nightly only, it might be worth having an enum like:

 ```rust
enum Search<'a> {
  Regexp(&'a str),
  Substring(&'a str),
}

impl <'a> Search<'a> {
  pub fn is_match(&self, other: &str) -> bool {}
}
```